packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -37,110 +37,97 @@ type RouterAction =
 };
 
-const MAX_HISTORY_LENGTH = 50;
-
 function addScreen( { screens }: RouterState, screen: Screen ) {
+	if ( screens.some( ( s ) => s.path === screen.path ) ) {
+		// eslint-disable-next-line no-console
+		console.warn(
+			`Navigator: a screen with path ${ screen.path } already exists.
+The screen with id ${ screen.id } will not be added.`
+		);
+		return screens;
+	}
 	return [ ...screens, screen ];
 }
 
 function removeScreen( { screens }: RouterState, screen: Screen ) {
 	return screens.filter( ( s ) => s.id !== screen.id );
 }
 
-function goBack( { locationHistory }: RouterState ) {
-	if ( locationHistory.length <= 1 ) {
-		return locationHistory;
-	}
-	return [
-		...locationHistory.slice( 0, -2 ),
-		{
-			...locationHistory[ locationHistory.length - 2 ],
-			isBack: true,
-			hasRestoredFocus: false,
-		},
-	];
-}
-
 function goTo(
 	state: RouterState,
 	path: string,
 	options: NavigateOptions = {}
 ) {
-	const { locationHistory } = state;
+	const { currentLocation, focusSelectors } = state;
+
 	const {
-		focusTargetSelector,
+		// Default assignments
 		isBack = false,
 		skipFocus = false,
-		replace = false,
+		// Extract to avoid forwarding
+		replace,
+		focusTargetSelector,
+		// Rest
 		...restOptions
 	} = options;
 
-	const isNavigatingToSamePath =
-		locationHistory.length > 0 &&
-		locationHistory[ locationHistory.length - 1 ].path === path;
-	if ( isNavigatingToSamePath ) {
-		return locationHistory;
-	}
-
-	const isNavigatingToPreviousPath =
-		isBack &&
-		locationHistory.length > 1 &&
-		locationHistory[ locationHistory.length - 2 ].path === path;
-
-	if ( isNavigatingToPreviousPath ) {
-		return goBack( state );
+	if ( currentLocation?.path === path ) {
+		return { currentLocation, focusSelectors };
 	}
 
-	const newLocation = {
-		...restOptions,
-		path,
-		isBack,
-		hasRestoredFocus: false,
-		skipFocus,
-	};
+	let focusSelectorsCopy;
 
-	if ( locationHistory.length === 0 ) {
-		return replace ? [] : [ newLocation ];
+	// Set a focus selector that will be used when navigating
+	// back to the current location.
+	if ( focusTargetSelector && currentLocation?.path ) {
+		if ( ! focusSelectorsCopy ) {
+			focusSelectorsCopy = new Map( state.focusSelectors );
+		}
+		focusSelectorsCopy.set( currentLocation.path, focusTargetSelector );
 	}
 
-	const newLocationHistory = locationHistory.slice(
-		locationHistory.length > MAX_HISTORY_LENGTH - 1 ? 1 : 0,
-		-1
-	);
-
-	if ( ! replace ) {
-		newLocationHistory.push(
-			// Assign `focusTargetSelector` to the previous location in history
-			// (the one we just navigated from).
-			{
-				...locationHistory[ locationHistory.length - 1 ],
-				focusTargetSelector,
-			}
-		);
+	// Get the focus selector for the new location.
+	let currentFocusSelector;
+	if ( isBack ) {
+		if ( ! focusSelectorsCopy ) {
+			focusSelectorsCopy = new Map( state.focusSelectors );
+		}
+		currentFocusSelector = focusSelectorsCopy.get( path );
+		focusSelectorsCopy.delete( path );
```

Comment:
Here I would enhance the logic for the case when we navigate to the `path` in a `isBack = false` way, and it has a `focusSelector` stored. Then we don't want to focus (that's implemented currently) and also to delete the stored `focusSelector` because it's no longer relevant.

Also, a little optimization opportunity: create the copy only when you really found a selector to be deleted. Currently, if `.get( path )` is `null`, you created the copy for nothing, it won't be mutated.